A32-xds-circuit-breaking.md

```diff
@@ -0,0 +1,162 @@
+messages, where each message specifies the parameters for a particular 
+RoutingPriority. gRPC will look only at the first entry in the list for 
+priority [`DEFAULT`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/base.proto#enum-core-routingpriority).
+
+Of the parameters in the `Thresholds` message, gRPC will support only
+`max_requests`, which sets the maximum number of requests that can be in flight
+to the cluster at any given time.  None of the other parameters are applicable
+to gRPC; for details, see [Other parameters considered](#other-parameters-considered).
+The design and implementation parts of this doc will mainly focus on the
+application of the `max_requests` circuit breaking parameter in gRPC.
+
+To match Envoy's implementation, circuit breaking parameters will always be 
+enforced with certain defaults, even if the gRPC client receives no applicable 
+circuit breakers. Users can effectively turn circuit breaking off by setting
+limits to very high values, for example, to 
+`std::numeric_limits<uint32_t>::max()`.
+
+## Detailed design
+
+With the introduction of [gRFC A28: xDS Traffic Splitting and Routing](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md), 
+it became possible for there to be more than one CDS LB policy instance for a 
+given cluster (e.g., if one route pointed to a given cluster and a different 
+route split traffic between that cluster and other clusters).  However, after 
+implementing the changes described in [gRPC A31: xDS RouteActions Support](https://github.com/grpc/proposal/pull/192),
+there will be only one CDS LB policy instance for a given cluster at any time
+in the LB policy tree. Since each CDS LB policy creates exactly one child EDS
+LB policy for endpoint discovery, both of them are part of the load balancing
+logic for requests sent to the specific cluster. Circuit breakers can be 
+implemented in either policy. Since cluster load assignment drops are handled 
+in EDS LB policy, it makes sense to handle circuit breaking drops in the same
+place.
+
+Each CDS LB policy will receive a configured limit for the maximum number of 
+outstanding requests to hosts in the cluster that this policy is load balancing
+for from the XdsClient cluster watch interface. A default value of 1024 is used
+if no such limit is specified by the management server. This limit will be 
+included in the LB config passed to its child EDS LB policy. The EDS LB config 
+will become
+
+```proto
+message EdsLoadBalancingPolicyConfig {
+  string cluster = 1;
+  string eds_service_name = 2;
+  google.protobuf.String lrs_load_reporting_server_name = 3;
+  repeated LoadBalancingConfig locality_picking_policy = 4;
+  repeated LoadBalancingConfig endpoint_picking_policy = 5;
+  
+  // Maximum number of outstanding requests can be made to the upstream cluster.
+  int32 max_concurrent_requests = 6;
```

Comment:
xDS uses `google.protobuf.UInt32Value` for the type of this field.  We should probably support the full unsigned integer range for compatibility.

We also need to document the default here, and we need to document how to disable this completely.  Envoy seems to disable it by setting it to the max value, which seems a little odd -- maybe we should use 0 for that instead?

Reply:
Envoy just uses a large number, so large that realistically not possible to achieve, to _effectively_ disable circuit breaking. This sounds legit and I don't think it's odd. One good thing for this is that implementations do not need to have special cases for the value. The value being 0? I guess a user will never set 0 for circuit breaking configurations if its usage is not specially noted, so a design doesn't need to focus on such extreme details and leave them as implementation flexibilities.

Reply:
I think you want to be consistent with Envoy here, control planes shouldn't have to think about two different values for disabling max concurrent requests.

Reply:
I absolutely agree that gRPC should not use a different "disable" value in the xDS resource than Envoy does, because we don't want control planes to have to send different data to different clients.  My question here was about how we represent this internally (the config passed from the CDS LB policy to the EDS LB policy), which could be different from how it's represented in xDS.  But I do also see the argument that using the same representation here makes sense, for consistency.

That having been said, I guess it's also worth asking whether we should allow using a 0 value to disable the limit even in xDS.  Using the maximum allowed value for a uint32 seems sub-optimal for a few reasons:

- It's fairly cumbersome for humans to populate, because people have to figure out what value to use.  (I know I never remember what MAX_UINT32 is; I always have to look it up or calculate it.)
- Do we actually know for certain that there will never be MAX_UINT32 concurrent requests in a given cluster?  The cluster could have a really large number of endpoints, and on a busy proxy on a really beefy machine, there really isn't any predefined upper bound here.  It seems sort of arbitrary to impose a limit just due to not having another way to disable this.
- From a performance perspective, setting the limit really high isn't really the same thing as turning off tracking completely, because there's still some overhead involved.  Ideally there should be a way to say "don't even bother tracking this, because I don't want the overhead".  (I suppose an implementation could special-case MAX_UINT32 to provide that behavior, but that seems a little hacky too.)

For all of these reasons, I am wondering if we should change xDS to allow using 0 to disable these circuit breakers.  That seems safe enough, since I can't think of any legitimate use-cases where one would actually want to set the limit to 0, since that would basically disallow all requests.

Harvey, thoughts...?

Reply:
I guess 0 might be used today by folks who want to disable all connections, but there's more direct ways to do that. Technically this would be a breaking API change.

Reply:
Yeah.  Okay, I guess let's not mess with this for now.  If we change it later, we can make that change in both Envoy and gRPC at that point.